testcases.py

```diff
@@ -736,6 +743,97 @@ def scenario() -> str:
         return "corrupt-rate --delay=15ms --bandwidth=10Mbps --queue=25 --rate_to_server=2 --rate_to_client=2"
 
 
+class TestCaseECN(TestCaseHandshake):
+    @staticmethod
+    def name():
+        return "ecn"
+
+    @staticmethod
+    def abbreviation():
+        return "E"
+
+    @staticmethod
+    def testname(p: Perspective):
+        return "handshake"
+
+    def count_ecn(self, tr):
+        ecn = [0] * (max(ECN) + 1)
+        for p in tr:
+            e = int(getattr(p["ip"], "dsfield.ecn"))
+            ecn[e] += 1
+        for e in ECN:
+            logging.debug("%s %d", e, ecn[e])
+        return ecn
+
+    def check_ecn_any(self, e):
+        return e[ECN.ECT0] != 0 or e[ECN.ECT1] != 0
+
+    def check_ecn_marks(self, e):
+        return (
+            e[ECN.NONE] == 0
+            and e[ECN.CE] == 0
+            and ((e[ECN.ECT0] == 0) != (e[ECN.ECT1] == 0))
+        )
+
+    def check_ack_ecn(self, tr):
+        # NOTE: We only check whether the trace contains any ACK-ECN information, not whether it is valid
+        for p in tr:
+            if hasattr(p["quic"], "ack.ect0_count"):
```

Comment:
Why are we only ECT0 here? ECT1 would be valid too, wouldn't it?

Reply:
This just checks if the attribute is there, which means there is an ACK-ECN frame. It seemed superfluous to check for the other ECN counts, as they will also be there in that case?

Reply:
What about implementations that mix up ECT0 and ECT1?

Reply:
I'm not looking at the value of the counter here, I just check whether the attribute exists, which I take as an indication that an ACK-ECN was sent.